### PR TITLE
stochastic beam search

### PR DESCRIPTION
**Pull request by qingerVT**
Summary:
Stochastic beam search aims to find top-k beams w/o replacement. They modify beam search at each time step using Gumbel-max trick but can guarantee beams are approximately optimal in global (complete sentences) in linear complexity. The generated translations are more diverse than BS, DBS, etc, and they show that this is helpful to estimate stats involving expectation over beams. We try to enable this in order to improve our RL training.

This paper won the runner-up reward at ICML'19 https://arxiv.org/abs/1903.06059
The algorithm is implemented in fairseq and the code is available online https://github.com/wouterkool/stochastic-beam-search

Differential Revision: D15871429

### CODE REVIEW

**Comment by myleott** [comment]
@wouterkool, would you be able to help review this? This PR is trying to simplify things a bit, and may also deviate slightly from the original paper; e.g., applying a previous fix to temperature (96ac28d) so that it operates over logits instead of logprobs.

Also, we'd love to mention this paper in the README. Would that be okay with you?

File: fairseq/sequence_generator.py [hunk 1]
```diff
@@ -58,9 +60,12 @@ def __init__(
                 (default: False)
             sampling_topk (int, optional): only sample among the top-k choices
                 at each step (default: -1)
-            temperature (float, optional): temperature, where values
-                >1.0 produce more uniform samples and values <1.0 produce
-                sharper samples (default: 1.0)
+            stochastic_beam_search: with --no-early-stop and --unnormalized to guarantee
```
[the remark this answers is not in the document]
Comment (qingerVT) [reply fairseq/sequence_generator.py]: yeah, good idea!

File: fairseq/sequence_generator.py [hunk 2]
```diff
@@ -153,6 +163,8 @@ def generate(
         # initialize buffers
         scores = src_tokens.new(bsz * beam_size, max_len + 1).float().fill_(0)
         scores_buf = scores.clone()
+        log_ps_t = src_tokens.data.new(bsz * beam_size, max_len + 1).float().fill_(0)
```
[the remark this answers is not in the document]
Comment (qingerVT) [reply fairseq/sequence_generator.py]: yeah, log_ps_t is only used for stochastic beam search, and I set it as None by default, i.e., step(, log_ps_t=None). I thought this for a while, and decided to keep it there since I don't want to add if, else when calling search.step() :-)

File: fairseq/search.py [hunk 3]
```diff
@@ -145,7 +199,7 @@ def step(self, step, lprobs, scores):
             else:
                 lprobs_g = lprobs_g.contiguous()
 
-            scores_buf, indices_buf, beams_buf = self.beam.step(step, lprobs_g, scores_g)
+            scores_buf, _, _, indices_buf, beams_buf = self.beam.step(step, lprobs_g, scores_g)
```
[the remark this answers is not in the document]
Comment (qingerVT) [reply fairseq/search.py]: This file is outdated. The latest one only has a single _,

**Comment by qingerVT** [comment]
Hey Wouter,

Thanks!! :-)

> * When sampling without replacement you obtain results _in an order_. To get an equivalent result, the final beam should be sorted decreasingly according to the perturbed log-probability (`scores` in the implementation). Does the current implementation do that or are shorter sentences finalized earlier? It depends on the use case whether this is relevant, and the user can also do the sorting manually but maybe it would be nice to do this by default/provide it as an option.

What does this mean? We should sort "score" decreasingly and then choose top-K? Could you please elaborate a bit?

> * In general I am not sure how 'scores' is defined/should be defined in fairseq with stochastic beam search (with deterministic beam search, both are equivalent):
>   
>   * As the selection for the beam search: then for SBS I think the score should be the perturbed logprob (as it is now)

  Yes, this is what it is now. When using vanilla or deterministic beam search, they are equivalent.

>   * As log-probability of partial sequence: then we should have a separate variable for the perturbation and maybe rename 'score' to logprob to avoid confusion?

So, you mean we need to have two variables: one w/o perturbation and one w/ perturbation? Then, set "score" to logprob, i.e., w/o perturbation, as it was in beam search (Fairseq), and using the variable w/ perturbation for ranking in SBS?

To clarify a bit. The variable "score" now is w/ perturbation. For deterministic beam search, it is equivalent to w/o perturbation, e.g, no changes here compared with previous code. For SBS, it is the score we need for ranking. 

I agree that having separate variables are clear, but it requires more memory and make inference slow in general. We hope to simplify our code as much as possible. 

Not sure I totally understand your concern, more explainable definitely helpful! Thanks!! 

Best,
Qing

**Comment by wouterkool** [comment]
Hi Qing,

I am very sorry, this one got out of my view. As I get quite some interest in Stochastic Beam Search, it would be really nice if we can still merge this into fairseq master! I'll answer your questions below. Let me know if there is anything I can help with.

Summarizing:
* I think we should sort the final results in decreasing order based on perturbed log-probability (or reason that the result is already sorted)
* We should consider adding documentation/comments on the definition of 'score' for stochastic beam search (i.e. with or without perturbation), especially in the final results returned in `get_hypo()`, where maybe we want to add 'logprob' and 'perturbed_logprob' which is less ambiguous.


> Hey Wouter,
> 
> Thanks!! :-)
> 
> > * When sampling without replacement you obtain results _in an order_. To get an equivalent result, the final beam should be sorted decreasingly according to the perturbed log-probability (`scores` in the implementation). Does the current implementation do that or are shorter sentences finalized earlier? It depends on the use case whether this is relevant, and the user can also do the sorting manually but maybe it would be nice to do this by default/provide it as an option.
> 
> What does this mean? We should sort "score" decreasingly and then choose top-K? Could you please elaborate a bit?

When sampling sequentially without replacement (for example by rejecting duplicates of samples with replacement), the results are obtained in a certain *order*. It depends on the application whether you care about the order or not, but if you want to interpret the result of Stochastic Beam Search as equivalent to sequential sampling without replacement, the final sequences should be sorted in descending order based on their perturbed log-probability.

So my question is, *is the final result sorted on perturbed log-probability*?

If it is the top k of the last step, and the top k operations returns results in order, then it should be so, but if I have understood fairseq's beam search implementation correctly, shorter sequences are finished earlier, which may affect the final order of results. In this case, we should *add an additional step to sort* the final results based on perturbed log-probability (or make that an option if this step is expensive but I think it would be good to default it to being sorted).

For example, if I have two partial sequences with perturbed log-probabilities [10, 8], then if the one with 8 is complete and directly finalized/added to a list with results, then later the partial sequence with perturbed log-probability 10 will yield a complete sequence with perturbed log-probability 10 and after finalizing we have two results with perturbed log-probabilities [8, 10], so *not* in decreasing order.

B.t.w. this is also the reason why early stopping is not compatible with stochastic beam search, since if the partial sequence with perturbed log-probability 10 gets two completions with perturbed log-probabilities 10 and 9, then the one with 8 should not be in the result, but it will be if we use early stopping.

> 
> > * In general I am not sure how 'scores' is defined/should be defined in fairseq with stochastic beam search (with deterministic beam search, both are equivalent):
> >   
> >   * As the selection for the beam search: then for SBS I think the score should be the perturbed logprob (as it is now)
> 
> Yes, this is what it is now. When using vanilla or deterministic beam search, they are equivalent.
> 
> > * As log-probability of partial sequence: then we should have a separate variable for the perturbation and maybe rename 'score' to logprob to avoid confusion?
> 
> So, you mean we need to have two variables: one w/o perturbation and one w/ perturbation? Then, set "score" to logprob, i.e., w/o perturbation, as it was in beam search (Fairseq), and using the variable w/ perturbation for ranking in SBS?
> 
> To clarify a bit. The variable "score" now is w/ perturbation. For deterministic beam search, it is equivalent to w/o perturbation, e.g, no changes here compared with previous code. Btw, I got the same bleu score before/after changes. For SBS, it is the score we need for ranking.
> 
> I agree that having separate variables are clear, but it requires more memory and make inference slow in general. We hope to simplify our code as much as possible.
> 
> Not sure whether I totally understand your concerns or not, please elaborate a bit, that would be very helpful! Thanks!!

I think it is good to decide if in the context of Stochastic Beam Search, 'score' should refer to log-probability or perturbed log-probability, and make sure this is clear in the code, as well as to users only dealing with the final set of output hypotheses. I am fine with both options, but I can imagine there can be confusion for a user who assumes hypothesis['score'] corresponds to 'log-probability' in the final hypothesis, which is not the case when using Stochastic Beam Search.

I think it would be worth considering to add 'logprob' and 'perturbed_logprob' (only for stochastic beam search) to the outputs in `get_hypo()`, which are less ambiguous, and consider dropping 'score' such that it cannot silently fail if people use Stochastic Beam Search and rely on score being the log-probability. See https://github.com/pytorch/fairseq/blob/2ec462262f91c30f12347e48760ac40eadc2cc31/fairseq/sequence_generator.py#L276 

> 
> Best,
> Qing

Best,

Wouter
